src/useClientRequest.js

```diff
@@ -2,104 +2,101 @@ const React = require('react');
 const ClientContext = require('./ClientContext');
 
 const actionTypes = {
-  REQUEST_LOADING: 'REQUEST_LOADING',
-  REQUEST_SUCCESS: 'REQUEST_SUCCESS',
-  REQUEST_FAILURE: 'REQUEST_FAILURE'
+  LOADING: 'LOADING',
+  CACHE_HIT: 'CACHE_HIT',
+  REQUEST_RESULT: 'REQUEST_RESULT'
 };
 
-function getInitialState(data) {
-  return {
-    data,
-    error: null,
-    loading: !data
-  };
-}
-
 function reducer(state, action) {
   switch (action.type) {
-    case actionTypes.REQUEST_LOADING:
+    case actionTypes.LOADING:
       return {
         ...state,
         loading: true
       };
-    case actionTypes.REQUEST_SUCCESS:
+    case actionTypes.CACHE_HIT:
       return {
-        ...state,
-        loading: false,
-        data: action.data
+        ...action.result,
+        cacheHit: true,
+        loading: false
       };
-    case actionTypes.REQUEST_FAILURE:
+    case actionTypes.REQUEST_RESULT:
       return {
-        ...state,
-        loading: false,
-        error: action.error
+        ...action.result,
+        cacheHit: false,
+        loading: false
       };
     default:
       return state;
   }
 }
 
-function useClientRequest(query, opts = {}) {
+/*
+  options include:
+
+  opts.variables: Object
+  opts.operationName: String
+  opts.fetchOptionsOverrides: Object
+  opts.skipCache: Boolean
+*/
+function useClientRequest(query, initialOpts = {}) {
   const client = React.useContext(ClientContext);
-  const { useCache } = opts;
-  const cacheKeyObject = {
+  const operation = {
     query,
-    ...opts
+    variables: initialOpts.variables,
+    operationName: initialOpts.operationName
   };
 
-  const initialState = getInitialState(getCacheHit(cacheKeyObject));
-  const [state, dispatch] = React.useReducer(reducer, initialState);
-
-  function getCacheHit(key) {
-    return useCache && !opts.skipCache && client.cache
-      ? client.cache.get(key) || null
-      : null;
-  }
+  const cacheKey = client.getCacheKey(operation, initialOpts);
+  const intialCacheHit = initialOpts.skipCache
+    ? null
+    : client.cache.get(cacheKey);
```

Comment:
Do we need to make `cache` a required option to be passed in? Or should we provide a default cache, or default to no cache?

Reply:
Cache is optional - this should check if client.cache exists.